### PR TITLE
1920020: show error message when setting invalid default_log_level as…

### PR DESCRIPTION
**Pull request by tlhmerry9**
… well as on every run until changed to a valid value

### CODE REVIEW

File: src/rhsm/config.py [hunk 1]
```diff
@@ -198,6 +201,15 @@ def set(self, section, name, value):
                 self.add_section(section)
             super(RhsmConfigParser, self).set(section, name, value)
 
+    def is_invalid_default_log_level(self, value):
+        if value not in ['CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG', 'NOSET']:
+            print("Invalid Log Level: {lvl}, setting to INFO for this run.".format(lvl=value), file=sys.stderr)
+            print(
+                "Please use:  subscription-manager config --logging.default_log_level=<Log Level> to set the default_log_level to a valid value.",
```
Comment (jirihnidek) [inline src/rhsm/config.py:208]: User usually does not know what valid values are, but it would be easy to print these valid values.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
When I set invalid value and I try to run some command then I get warning value twice as you can see:

```
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager role --show
Invalid Log Level: FOO, setting to INFO for this run.
Please use:  subscription-manager config --logging.default_log_level=<Log Level> to set the default_log_level to a valid value.
Invalid Log Level: FOO, setting to INFO for this run.
Please use:  subscription-manager config --logging.default_log_level=<Log Level> to set the default_log_level to a valid value.
Current Role: SP Starter
```

Please modify code to print this warning message only once in this case.

File: src/rhsm/config.py (hunk 1 above)
Comment (tlhmerry9) [reply src/rhsm/config.py:208]: added valid values to the message

File: src/rhsmlib/dbus/server.py [hunk 2]
```diff
@@ -31,16 +31,12 @@
...
 from subscription_manager import injection as inj
-from rhsm.logutil import init_logger
-
 
 log = logging.getLogger(__name__)
 
 parser = get_config_parser()
 conf = config.Config(parser)
 
-init_logger(parser)
```
Comment (jirihnidek) [inline src/rhsmlib/dbus/server.py:42]: Calling all of this here is wrong, but simple removing this code had bad side effect. We lost logging of rhsm.service. I would move it to some method (maybe `__init__()` of `Server` class).

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Hi, thanks for update. It's much more better now. :+1: Your change has one wrong side-effect that could be easily fixed and it when I had some wrong value of `logging.default_log_level` and I'm setting new right value, then I can still can see warning message (further calling of subscription-manager is without warning message):

```
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager config --logging.default_log_level=DEBUG
Invalid Log Level: foo, setting to INFO for this run.
Please use:  subscription-manager config --logging.default_log_level=<Log Level> to set the default_log_level to a valid value.
Valid Values: ['CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG', 'NOSET']
```

I guess that setting values and checking values is not in right order.

File: src/rhsm/config.py [hunk 3]
```diff
@@ -198,6 +201,17 @@ def set(self, section, name, value):
                 self.add_section(section)
             super(RhsmConfigParser, self).set(section, name, value)
 
+    def is_invalid_default_log_level(self, value):
```
Comment (ptoscano) [inline src/rhsm/config.py:204]: Minor readability note: I'd change this name into a positive name, e.g. `is_default_log_level_valid`. A negative name (is_invalid) makes it slightly harder to understand its effect when reading the code, twisting it even more when adding negations in conditions (`if not is_invalid():`).

Also, the "default" is unrelated here, as this function merely checks what is passed to it. Hence, as simpler `is_log_level_valid` will do.

File: src/rhsm/config.py [hunk 4]
```diff
@@ -198,6 +201,17 @@ def set(self, section, name, value):
...
+        valid = ['CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG', 'NOSET']
+        if value not in valid:
+            print("Invalid Log Level: {lvl}, setting to INFO for this run.".format(lvl=value), file=sys.stderr)
+            print(
+                "Please use:  subscription-manager config --logging.default_log_level=<Log Level> to set the default_log_level to a valid value.",
+                file=sys.stderr)
+            print("Valid Values: {valid}".format(valid=valid), file=sys.stderr)
```
[the remark this answers is not in the document]
Comment (ptoscano) [reply src/rhsm/config.py:211]: pity this code cannot use `subscription_manager.utils.friendly_join()` :-/

[the remark this answers is not in the document]
Comment (tlhmerry9) [reply src/rhsm/config.py:211]: updated to a string in main,

File: src/rhsm/config.py (hunk 3 above)
Comment (tlhmerry9) [reply src/rhsm/config.py:204]: I changed the name in main

File: src/rhsmlib/dbus/server.py (hunk 2 above)
Comment (tlhmerry9) [reply src/rhsmlib/dbus/server.py:42]: moved to the __init__() of Server class.

File: src/rhsm/config.py (hunk 3 above)
Comment (tlhmerry9) [reply src/rhsm/config.py]: changed the name to is_log_level_valid

**Comment by tlhmerry9** [comment]
> Hi, thanks for update. It's much more better now.  Your change has one wrong side-effect that could be easily fixed and it when I had some wrong value of `logging.default_log_level` and I'm setting new right value, then I can still can see warning message (further calling of subscription-manager is without warning message):
> 
> ```
> [root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager config --logging.default_log_level=DEBUG
> Invalid Log Level: foo, setting to INFO for this run.
> Please use:  subscription-manager config --logging.default_log_level=<Log Level> to set the default_log_level to a valid value.
> Valid Values: ['CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG', 'NOSET']
> ```
> 
> I guess that setting values and checking values is not in right order.

The reason this happens is that the first thing that is done is to read the config file for the default_log_level when the init_logger method is called.  This is before any value is attempted to be set.

**Comment by jirihnidek** [comment]
> > Hi, thanks for update. It's much more better now.  Your change has one wrong side-effect that could be easily fixed and it when I had some wrong value of `logging.default_log_level` and I'm setting new right value, then I can still can see warning message (further calling of subscription-manager is without warning message):
> > ```
> > [root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager config --logging.default_log_level=DEBUG
> > Invalid Log Level: foo, setting to INFO for this run.
> > Please use:  subscription-manager config --logging.default_log_level=<Log Level> to set the default_log_level to a valid value.
> > Valid Values: ['CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG', 'NOSET']
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I guess that setting values and checking values is not in right order.
> 
> The reason this happens is that the first thing that is done is to read the config file for the default_log_level when the init_logger method is called. This is before any value is attempted to be set.

Yeah, I suspected something like that. I would open another BZ for this case and I would fix it as part of another PR to keep this PR simple.

**Review by jirihnidek — APPROVED** [review]
ACK

**Pull request merged**